core/templates/dev/head/components/top_navigation_bar/TopNavigationBarDirective.js

```diff
@@ -95,13 +95,92 @@ oppia.directive('topNavigationBar', [function() {
           $scope.windowIsNarrow = windowDimensionsService.isWindowNarrow();
           $scope.$apply();
           // Close the sidebar, if necessary.
           SidebarStatusService.closeSidebar();
+          // Limit calls to handleOverflow() to one per 500ms to prevent flicker
+          $timeout.cancel(handleOverflowResizeTimer);
+
+          // If the window is resized larger, try displaying the hidden elements
+          if (initialWindowWidth < windowDimensionsService.getWidth()) {
+            $.each(hideNavElements, function(element, state) {
+              if (state === true) {
+                $('[translate=' + element + ']')
+                  .closest('div, li').css('display', 'block');
+                hideNavElements[element] = false;
+              }
+            });
+          }
+          initialWindowWidth = windowDimensionsService.getWidth();
+          handleOverflowResizeTimer = $timeout(handleOverflow, 500);
         });
 
+        var handleOverflow = function() {
+          if (windowDimensionsService.getWidth() < 768) {
+            return false;
+          }
+
+          var navReady = true;
+          // Wait until i18n is completed.
+          $('.oppia-navbar-tabs a[translate], ' +
+            '.oppia-navbar-tabs span[translate]').each(function(i, element) {
+              if (element.innerText.length === 0) {
+                $timeout(handleOverflow, 100);
+                navReady = false;
+                return false;
+              }
+            });
+          if (!navReady) {
+            return false;
+          }
+
+          var navWidth = 0;
+          $('ul.oppia-navbar-tabs').children().each(function(i, element) {
+            // Some nav elements have invalid widths from floating
+            // Use widths of their first-level child elements instead
+            if (element.clientWidth === 0 || element.clientWidth >= 176) {
+              $(element).children().each(function(i, element) {
+                navWidth += 5;
+                navWidth += element.clientWidth;
+              });
+            } else {
+              navWidth += 5;
+              navWidth += element.clientWidth;
+            }
+          });
+
+          if ($('ul.nav.oppia-navbar-profile').length > 0) {
+            navWidth += $('ul.nav.oppia-navbar-profile').width();
+          }
+
+          $('ul.nav.oppia-navbar-tabs').css('min-width', navWidth);
+
+          if ($('div.collapse.navbar-collapse').height() > 60) {
+            $.each(hideNavElements, function(element, state) {
+              if (state === false) {
+                $('[translate=' + element + ']')
+                  .closest('div, li').css('display', 'None');
+                hideNavElements[element] = true;
+                $timeout(handleOverflow, 10);
+                return false;
+              }
+            });
+          }
+        };
+
+        // For Chrome, timeout 0 appears to run after i18n.
```

Comment:
I measured the height of the navbar to be 59.xxx when not overflowed (in inspector), so I used a check for >60 in the function.

If it's over 60, I go through each of the elements to hide: if the element is already hidden, go to the next. If the element is not hidden, hide it and then call the handleOverflow *to be renamed function again after 10ms (to give the browser time to render the css change). Only one element will be hidden per call of the function.